app/modules/cdr/endpoints_cdr.py

```diff
@@ -0,0 +1,2816 @@
+                membership_id=existing_membership.id,
+                membership=schemas_cdr.MembershipEdit(
+                    end_date=existing_membership.end_date
+                    + product_variant.related_membership_added_duration,
+                ),
+            )
+        else:
+            added_membership = models_core.CoreAssociationMembership(
+                id=uuid4(),
+                user_id=user_id,
+                membership=product.related_membership,
+                start_date=date(datetime.now(tz=UTC).date().year, 9, 1),
+                end_date=date(datetime.now(tz=UTC).date().year, 9, 1)
+                + product_variant.related_membership_added_duration,
+            )
+            cruds_cdr.create_membership(db=db, membership=added_membership)
+
+
+@module.router.patch(
+    "/cdr/users/{user_id}/purchases/{product_variant_id}/validated/",
+    status_code=204,
+)
+async def mark_purchase_as_validated(
+    user_id: str,
+    product_variant_id: UUID,
+    validated: bool,
+    db: AsyncSession = Depends(get_db),
+    user: models_core.CoreUser = Depends(is_user_a_member_of(GroupType.admin_cdr)),
+):
+    """
+    Validate a purchase.
+
+    **User must be CDR Admin to use this endpoint**
+    """
+    db_purchase = await cruds_cdr.get_purchase_by_id(
+        db=db,
+        user_id=user_id,
+        product_variant_id=product_variant_id,
+    )
+    if not db_purchase:
+        raise HTTPException(
+            status_code=404,
+            detail="Invalid purchase",
+        )
+
+    product_variant = await cruds_cdr.get_product_variant_by_id(
+        db=db,
+        variant_id=product_variant_id,
+    )
+    if not product_variant:
+        raise HTTPException(
+            status_code=404,
+            detail="Invalid product_variant_id",
+        )
+    product = await cruds_cdr.get_product_by_id(
+        db=db,
+        product_id=product_variant.product_id,
+    )
+    if not product:
+        raise HTTPException(
+            status_code=404,
+            detail="Invalid product.",
+        )
+    if validated:
+        memberships = await cruds_cdr.get_actual_memberships_by_user_id(
+            db=db,
+            user_id=user_id,
+        )
+        for product_constraint in product.product_constraints:
+            purchases = await cruds_cdr.get_purchases_by_ids(
+                db=db,
+                user_id=user_id,
+                product_variant_id=[
+                    variant.id for variant in product_constraint.variants
+                ],
+            )
+            if not purchases:
+                if product_constraint.related_membership:
+                    if product_constraint.related_membership not in [
+                        m.membership for m in memberships
+                    ]:
+                        raise HTTPException(
+                            status_code=403,
+                            detail=f"Product constraint {product_constraint} not satisfied.",
```

Comment:
We should return the name of the product, instead of the whole Python object, to get a more explicit error
<img width="413" alt="Capture d’écran 2024-08-20 à 17 35 02" src="https://github.com/user-attachments/assets/0a83a461-1db1-41a2-bc8f-08b9183a5db7">